### PR TITLE
Allow the base route to be altered

### LINKED ISSUE #202: Unable to configure and use baseurl in rails application
Hello,
I have configured baseurl in my application js file like.
 `
   MessageBus.baseUrl = 'fa/';
    MessageBus.start();
    MessageBus.subscribe("/notification_channel_" + window.wsConfig.sessionId, function(data){
        console.log(data);
    }
`
I keep getting following error in my Rails application
`ActionController::RoutingError` (No route matches [POST] `"/fa/message-bus/fc1684579dd54b9bb0a89fd8e4a3fa5d/poll.`

I tried exploring lib/message_bus/rack/middleware.rb file but could not find any code handling response with custom baseurl. Please help. Thanks

### PR DESCRIPTION
**Pull request by nikolai-b**
Resolves #202

Allows the route that message bus listens on to be configured

### CODE REVIEW

**Comment by SamSaffron** [comment]
I think this looks ok, can you add a note to release notes and readme about this?

File: lib/message_bus/rack/middleware.rb [hunk 1]
```diff
@@ -54,7 +54,7 @@ def stop_listener
   # Process an HTTP request from a subscriber client
   # @param [Rack::Request::Env] env the request environment
   def call(env)
-    return @app.call(env) unless env['PATH_INFO'] =~ /^\/message-bus\//
+    return @app.call(env) unless env['PATH_INFO'] =~ /^#{@bus.base_route}message-bus\//
```
Comment (SamSaffron) [inline lib/message_bus/rack/middleware.rb:57]: might as well memoize this regex

Comment (SamSaffron) [reply lib/message_bus/rack/middleware.rb:57]: I wonder do we even need a regex here ... why not just use start_with?

**Comment by nikolai-b** [comment]
Not sure why CI is failing, looks like it got stuck

File: lib/message_bus/rack/middleware.rb [hunk 2]
```diff
@@ -39,6 +39,7 @@ def initialize(app, config = {})
     @bus = config[:message_bus] || MessageBus
     @connection_manager = MessageBus::ConnectionManager.new(@bus)
     @started_listener = false
+    @client_id_regexp = Regexp.new(@bus.base_route + 'message-bus/(?<client_id>[[:alnum:]]+)')
```
Comment (SamSaffron) [inline lib/message_bus/rack/middleware.rb:42]: I think you need a regex escape here ... base route could be `test.it` ... then this would match `test7it`

Comment (nikolai-b) [reply lib/message_bus/rack/middleware.rb]: Ah, thanks.  I've changed it to using a substring as I was also thinking that client_id might have `-` or other characters.  I thought it was slower but looks like it is about the same (or faster even)

```ruby
base_route = "/base/route/";
path_env = "#{base_route}message-bus/clientId";

rgx = Regexp.new(base_route + 'message-bus/(?<client_id>[[:alnum:]]+)');

substr_proc = -> { path_env["#{base_route}message-bus/".length..-1].split("/")[0] };
regex_proc = -> { path_env.match(rgx) { |m| m[:client_id] } };
split_proc = -> { path_env.split("#{base_route}message-bus/")[1].split("/")[0] };

substr_proc.call # gives "clientId"
regex_proc.call # gives "clientId"
split_proc.call # gives "clientId"

Benchmark.ips { |x| x.report("substr") { substr_proc.call }; x.report("regexp") { regex_proc.call }; x.report("split") { split_proc.call } };
```

```
Warming up --------------------------------------
              substr    75.167k i/100ms
              regexp    67.942k i/100ms
               split    65.618k i/100ms
Calculating -------------------------------------
              substr      1.004M (± 8.0%) i/s -      5.036M in   5.048246s
              regexp    801.254k (±13.0%) i/s -      3.941M in   5.013270s
               split    799.437k (± 6.1%) i/s -      4.003M in   5.025920s
```

File: lib/message_bus/rack/middleware.rb [hunk 3]
```diff
@@ -54,7 +55,7 @@ def stop_listener
   # Process an HTTP request from a subscriber client
   # @param [Rack::Request::Env] env the request environment
   def call(env)
-    return @app.call(env) unless env['PATH_INFO'] =~ /^\/message-bus\//
+    return @app.call(env) unless env['PATH_INFO'].start_with? "#{@bus.base_route}message-bus/"
```
Comment (SamSaffron) [inline lib/message_bus/rack/middleware.rb:58]: I know I am being ultra stickler here, but might as well micro optimise this cause it is called on every request.

Maybe memoize this value at `@bus.base_path` then you just keep one of these strings around for the life of the process.

**Comment by SamSaffron** [comment]
OK lets give this a shot! thanks @nikolai-b !

**Pull request merged**